core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServices.java

```diff
@@ -326,6 +340,48 @@ private static ResponseConverter converter(Method method) {
     }
 
     /**
+     * Returns a decorator chain which is specified by {@link Decorate} annotations.
+     */
+    private static Function<Service<HttpRequest, HttpResponse>,
+            ? extends Service<HttpRequest, HttpResponse>> decorator(Method method) {
+
+        final Decorate decorate = method.getAnnotation(Decorate.class);
+        if (decorate == null) {
+            return Function.identity();
+        }
+
+        final Class<? extends DecoratingServiceFunction<HttpRequest, HttpResponse>>[] c = decorate.value();
+        if (c.length == 0) {
+            return Function.identity();
+        }
+
+        // Respect the order of decorators which is specified by a user. The first one is first applied.
+        Function<Service<HttpRequest, HttpResponse>,
+                ? extends Service<HttpRequest, HttpResponse>> decorator = newDecorator(c[c.length - 1]);
+        for (int i = c.length - 2; i >= 0; i--) {
+            decorator = decorator.andThen(newDecorator(c[i]));
+        }
+        return decorator;
+    }
+
+    /**
+     * Returns a new decorator which decorates a {@link Service} by {@link FunctionalDecoratingService}
+     * and the specified {@link DecoratingServiceFunction}.
+     */
+    private static Function<Service<HttpRequest, HttpResponse>,
+            ? extends Service<HttpRequest, HttpResponse>> newDecorator(
+            Class<? extends DecoratingServiceFunction<HttpRequest, HttpResponse>> clazz) {
+        return service -> new FunctionalDecoratingService<>(
+                service, DECORATING_SERVICE_FUNCTIONS.computeIfAbsent(clazz, newClazz -> {
+            try {
+                return newClazz.newInstance();
```

Comment:
- Perhaps we could try to call `setAccessible(true)` on the constructor?
- How about also trying to call `setAccessible(true)` on all annotated service methods, so a user can use an anonymous class like this:

  ```java
  sb.annotatedService(new Object() {
      // Fails even if hello() is public.
      @Get("/")
      public AggregatedHttpMessage hello() {
          return AggregatedHttpMessage.of(HttpStatus.OK, MediaType.PLAIN_TEXT_UTF_8, "Hello!");
      }
  });
  ```